### ISSUE TITLE
Outdated apt package and release tag

### ISSUE THREAD

**Issue by clydeshaffer** [cc65/cc65#2281]
I've had this come up as a friction point a couple of times where someone tried to build my code, got errors that I didn't, and it turned out the reason was that they installed cc65 with apt or they used the most recent build on the "releases" tab.

One example is that the signature of inflatemem() was changed and now causes an error by default when building code written for recent cc65 with the "latest release" of cc65.

Could the version number and release tag be bumped? Or is there a particular reason that it's been left alone for four years?

**Comment by mrdudz** [comment]
see https://github.com/cc65/wiki/wiki/Before-the-next-release

**Comment by jordigh** [comment]
Can we just do a release anyway? It doesn't matter that it's not perfect and that it has all of these outstanding issues. The Windows snapshot has become the de facto thing to install anyway for many people, so why not just make a release with whatever state things are in?

**Comment by mrdudz** [comment]
> What are the specific requirements, or in other words what is the work that you would otherwise do to "produce" a release?

I wouldn't even know. I guess people expect a package containing binaries, which they can install? And be able to download it from here?

> I'd take a try at the GHA but given there's already a button on GH to select a tag and create a release I'm not sure what else the GHA is supposed to do.

The github action should be triggered when a git tag was made, then compile all configurations, run the tests, produce windows and linux packages, and then publish them as a release on github.

For VICE the respective GHA looks like this: https://github.com/VICE-Team/svn-mirror/blob/main/.github/workflows/make-release.yml

**Comment by jordigh** [comment]
Totally down to write the github action. [That's what I do anyway for my day job.](https://github.com/gristlabs/grist-core/commit/6760416a249f77b57a54da410012ff7962c9e022)

I'll test on a fork and submit a PR this weekend.

**Comment by mrdudz** [comment]
@jordigh any results? it would be really nice to have, obviously :)

**Comment by gilmorem560** [comment]
+1 vote for this, imo whatever gets package managers out there to upgrade their stuff, they can build packages if they need to obviating the need for a release process to also include producing a bunch of binaries. WinNT folks are a bunch of babies who expect every vendor to hand them an exe so I could see a need to roll your own for Windows folks, but between package managers and Homebrew in the macOS world, people using real computers are covered without you having to make binaries.

**Comment by jordigh** [comment]
Oh wow, hah, I said I'd do this in January, eh?

I'll get back to this task.

**Comment by gilmorem560** [comment]
You're a hero don't feel bad. It's only on my radar recently due to a da65 gap I'm currently dealing with in v2.19. This probably hasn't been a bigger deal because the people who do need what has happened in cc65 since 2020 have just bumped to the current trunk. That may honestly bode well for cutting a new release sometime soon by hand but then since that has caught releases up to 2025 from 2020, I think folks could then be understanding of the time then being put in to create a new release engineering protocol. A v.2.20 should sate people for quite a while.

**Comment by GorillaSapiens** [comment]
see my comments in #1347.  we really should put a release together.  while it may not be everything you want right now, i am absolutely certain that some progress has been made in the last FIVE YEARS.

**Comment by krischik** [comment]
> should be as easy as tagging a release

Just do that: `git tag V2.19.1` and `git push --tags` at any release that is half decent.

**Comment by krischik** [comment]
> I guess people expect a package containing binaries, which they can install?

I don't. I'm fine with at source code zip I know will compile and run. Everything else is just nice to have.

**Comment by mrdudz** [comment]
> I'm fine with at source code zip I know will compile and run.

But the link to the tarball is right there. I don't get it.

**Comment by gilmorem560** [comment]
Yeah generating the tag will automatically post the zip and tarball of the source code, binaries and other artifacts are optional. Generally folks in the Linux and BSD ecosystems will probably be grabbing this from their package managers. For macOS there's Homebrew, leaving Windows as the only platform where you might consider rolling a binary distro for, dunno if this project has historically had third-parties that provide windows builds, but I've seen that sort of thing out there in the world too. If someone close to the project has a WinNT box, you can probably just grab the VS Build Tools to get at the compilers and such, you don't have to mess with the full bulk of Visual Studio. Of course there's also mingw and building a gcc cross compiler, you could probably finagle a build container like a Docker thing to just poop out WinNT builds. Either way, even in this year of 2025, the i686 target is your safest choice. WinNT-on-ARM people will just have to deal unless you really want to roll those binaries too.

**Comment by mrdudz** [comment]
We do have automatic builds (of the git head)

**Comment by krischik** [comment]
> > I'm fine with at source code zip I know will compile and run.
> 
> But the link to the tarball is right there. I don't get it.

But it's 5 years old: 

<img width="1686" alt="Image" src="https://github.com/user-attachments/assets/a6f782fd-ac2c-432a-b8f3-239708ab438c" />

And no, HEAD is a moving target. I can't use HEAD for MacPorts. I need a tag or a stable branch (for example using git-flow). You don't have either. And it's same for the 6502 projects i publish. What should I tell those who download which  CA65 to use? I need to put a version into the documentation.

Note that I use git-flow for all my projects - even when I'm the only developer. I create release braches for all of them, make final unit tests and then create a release. For the smaller project this a quarter hour work.

And I do this for projects so obscure I don't expect more then dozen people do ever download them.

**Comment by mrdudz** [comment]
Thats a bit more than "I'm fine with at source code zip I know will compile and run" though. (I am sure you can make a macports recipe for both git HEAD or a specific pinned version)

That said, i'd always tell people to download the snapshot. Its almost always the better choice :)

**Comment by krischik** [comment]
> Thats a bit more than "I'm fine with at source code zip I know will compile and run" though. (I am sure you can make a macports recipe for both git HEAD or a specific pinned version)

What could could do and what is a good idea are two different things. 

- HEAD — is an extremely bad idea as everybody would get a different version. An absolute support nightmare. Also MacPorts used sha256 to check source code integrity.
- pinned version — which version am supposed to pin? You could pin a version because you know what version is good. I can't.

But if you want to do it here go ahead: https://github.com/macports/macports-ports/blob/master/lang/cc65/Portfile

> That said, i'd always tell people to download the snapshot. Its almost always the better choice :)

That's is bad advice because there is no guarantee that and snapshot actually works. At any one moment you could do a breaking change and my projects won't compile any more. In fact it probably won't as I'm still use 2.9 and there is a high likelihood you did a breaking change already.

**Comment by krischik** [comment]
Quick check:

```
>brew info cc65
==> cc65: stable 2.19 (bottled), HEAD
6502 C compiler
https://cc65.github.io/cc65/
Installed
/usr/local/Cellar/cc65/2.19 (481 files, 31MB) *
  Poured from bottle on 2020-11-23 at 08:44:12
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/c/cc65.rb
License: Zlib
```

Brew ist also 2.9.

**Comment by GorillaSapiens** [comment]
*ugh* just found this in src/common/version.c

if we ever do convince PIC (People In Charge) to do a new release, it should really start with something higher than 3.48 and fix this mess once and for all.

<pre>
*
** Note: until 2.19 the __CC65__ macro was defined as (VER_MAJOR * 0x100) + VER_MINOR * 0x10
** which resulted in broken values starting at version 2.16 of the compiler:
**
** version  __CC65__    identifies as
**
** 2.0      0x0200      2.0
** 2.1      0x0210      2.16
** 2.2      0x0220      2.32
** 2.3      0x0230      2.48
** 2.4      0x0240      2.64
** 2.5      0x0250      2.80
** 2.6      0x0260      2.96
** 2.7      0x0270      2.112
** 2.8      0x0280      2.128
** 2.9      0x0290      2.144
** 2.10     0x02a0      2.160
** 2.11     0x02b0      2.176
** 2.12     0x02c0      2.192
** 2.13     0x02d0      2.208
** 2.14     0x02e0      2.224
** 2.15     0x02f0      2.240
** 2.16     0x0300      3.0
** 2.17     0x0310      3.16
** 2.18     0x0320      3.32
** 2.19     0x0330      3.48
** 2.19-git 0x0213      2.19
**
** to keep damage low(er), we should do the following:
**
** - bump to 3.1 (skip 3.0) before 2.32
** - bump to 4.0 before 3.16
**
** That way at least each value is unique, and checking compiler version(s) can
** still work to some degree, should it really be necessary.
**
** Some preprocessor kludges can be used to still check for greater or lesser
** versions - see the checkversion program in the samples directory.
**
*/
</pre>

**Comment by krischik** [comment]
> if we ever do convince PIC (People In Charge) to do a new release, it should really start with something higher than 3.48 and fix this mess once and for all.

What a mess. Did no one notice that this version scheme would break with 2.16? And why `VER_MINOR * 0x10`? That was unnecessary as there where never any bug fix releases.

Yes. a jump to `4.0 0x0400 4.00` is probably the best idea. But I fear the People In Charge don't care if there product is successful or not.

**Comment by GorillaSapiens** [comment]
see #2757

**Comment by mrdudz** [comment]
> Did no one notice that this version scheme would break with 2.16?

If i didnt put this comment there when i stumbled about (and fixed) it, noone would have noticed its broken until today. Its a non issue too - unless you are mixing future compilers with VERY OLD compilers in your dev environment - in which case you deserve all you get for it :=P

**Comment by GorillaSapiens** [comment]
You "fixed it" by going backwards? That's the opposite of a fix.

**Comment by mrdudz** [comment]
I am going to delete further rambling from you that has nothing to do with the ticket.

**Comment by GorillaSapiens** [comment]
*shrug* it's not like you're listening anyway

**Comment by clydeshaffer** [comment]
Yeah, I think it might be a good idea to remain respectful when this resource is maintained for free as essentially a hobby.

Back on topic, I do think it would be a good idea to every so often mark the current HEAD state with a version tag. It'd be really helpful for projects that depend on the compiler, even if the tagged version doesn't have any auto builds attached. I'd rather have some of the benefits of version tagging than none, while waiting for it to be made perfect.

**Comment by krischik** [comment]
> Back on topic, I do think it would be a good idea to every so often mark the current HEAD state with a version tag.

Even making a tag every 6 month on the current version would be helpful for package maintainer.

**Comment by mrdudz** [comment]
> Even making a tag every 6 month on the current version would be helpful for package maintainer.

That way you'll be better off just using master - you'll get random versions that may have random bugs, and not only that, you'll be stuck with them for 6 months and get no support other than "use head and come back". I doubt that is a good idea at all.

Anyway, all this doesn't matter. The key to get releases is:

i am not going to do them manually, everything must happen automatically when a release is tagged. And that includes:

- building the binaries, docs etc 
- run all the tests
- produce a zip with windows binaries, docs
- update the website and whatever else, ie copy some files around to where they belong
(plus there needs to be a howto that tells how it all works, how to test it in own repo - and most importantly how to revert a release should something went wrong).

The puzzle pieces are all there already too - it just needs someone who knows this YML stuff and GHA. @dqh-au is busy with other stuff unfortunately. @jordigh offered to help for that matter :)

That said, *right now* i'd not make a release anyway, due to the activity and some recent changes. I'd always wait for a few weeks with no activity before a release. With VICE we usually "freeze" the code about 6 weeks before release, and then only bug fixes and documentation will be made until release.

**Comment by clydeshaffer** [comment]
> That way you'll be better off just using master - you'll get random versions that may have random bugs, and not only that, you'll be stuck with them for 6 months and get no support other than "use head and come back". I doubt that is a good idea at all.

If it's a set of bugs that can be worked around then at least they're a *consistent* set of bugs that can be paired to a codebase. Code that works with head today might not work with head in six months. I'm not trying to antagonize, I'm just having trouble understanding the resistance to doing any tags without all these other deliverables you seem convinced are required to do any tags.

You don't want to manually build binaries, run tests, produce zips, and update the website; and nobody is asking you to do those. So why are tags blocked on those?

**Comment by mrdudz** [comment]
Because anything that isn't properly packaged and linked from the proper places is not a release to me. If its not even that - then its equally good to just use head. And then you should use just that. I still don't get why you don't just pick a version yourself for that matter.

So please lets stop this circle. @jordigh can hopefully make this GHA work, and then all is fine. (Or someone else could, obviously)

**Comment by GorillaSapiens** [comment]
> You don't want to manually build binaries, run tests, produce zips, and update the website; and nobody is asking you to do those. So why are tags blocked on those?

zips and tarballs created by prior tagged releases do not have executables.  at least not the ones on github.  they're all just packaged up source code.

tests get run when a PR is created.  i don't think tagging a revision really needs to trigger any more tests.

can't speak to the website.

surprisingly (to me anyway) the "getting started" page on the website has a link to "current cc65 snapshot" that takes you to a sourceforge (!?) page with zip files containing win32 and win64 binaries.  even more surprising, the binaries are built from current code.  no clue how that happens.

_usual bickering removed. bye_

**Comment by mrdudz** [comment]
> zips and tarballs created by prior tagged releases do not have executables.

yes, that too. Windows releases require extra actions. And Windows is probably 95% of our user base.

> no clue how that happens.

And that is simply the GHA - which uploads the produced files somewhere. This couldn't be done directly to github for some reason. At least years ago when @oliverschmidt set up those things - perhaps it is doable these days. Doesn't really matter, THIS part of the deal is working

**Comment by krischik** [comment]
> you'll be stuck with them for 6 months

Currently the user are stuck with a 5 year old version. So the user has no new features and no bug fixes.

**Comment by krischik** [comment]
> then its equally good to just use head.

HEAD is never better because it's random. Especially since this project isn't using git-flow and doesn't have a **develop** branch. As such **master** is not even guaranteed to compile as two developer might be merging their work right now. Or developer accidentally made a bad commit.

Note that I use git-flow even on one-man-projects because it allows me to structure my work much better then working on a single branch.

**Comment by groessler** [comment]
Why cannot you use the git hash which is current as your cc65 version for the package? And you can use a newer git hash every 6 months or year.
Why do you need a tag?

**Comment by krischik** [comment]
> Why cannot you use the git hash which is current as your cc65 version for the package? And you can use a newer git hash every 6 months or year. Why do you need a tag?

I see you have never worked with a package manager. Package manager like MacPorts are optimised to work with a tar or zip files containing the source code and absolutely needs a version number. Package are not ment to be used with the repository directly.

MacPorts for example has [mirror sites](https://trac.macports.org/wiki/Mirrors) to keep copies of the source code in case the master site goes down.

A tag will provide both: a source zip file and a version number.

You might be able to tweak MacPorts to use a git hash but that is significantly more work. If you want to do it go ahead.

**Comment by mrdudz** [comment]
I am using packages that either use "head" or some pinned version all the time - official packages from my distro that is. Sounds more of a problem in macports than package managers in general. (And eg you can install the svn head of VICE using either brew or macports, i forgot)

In any case, all of this has been said, we are already going in circles. Lets see if @jordigh comes up with a GHA

**Comment by gilmorem560** [comment]
That is a good point, it is ultimately up to package repos to decide their practices regarding using releases vs commit heads vs whatever else. Really in my mind the urgency on a new release to address this depends on how likely people are to be using post-2.19 features in active development they expect other people to be able to build independently. Since currently you can't predict whether a given user will have anything newer, the safe bet as a developer is to preclude any post-2.19 features. Once a 2.20 or whatever it would be drops, especially once adopted by distros, then it is safe to use post-2.19 features for collaborative work.

Either way, there are multiple ways to solve this problem, one being a release, another being restraint on the part of developers when they know someone may need to use the current release on their stuff. If you're working on a solo project, then all bets are off and yeah, you can just go roll a trunk build.

**Comment by oliverschmidt** [comment]
The FujiNet project is a pretty heavy user of cc65. There are quite some apps. C, asm and mixed. Atari, Apple II and C64. For them, the current release is completely worthless. The clear instruction for all devs is to get the cc65 HEAD, build it and go from there.

**Comment by mrdudz** [comment]
> The FujiNet project is a pretty heavy user of cc65. There are quite some apps. C, asm and mixed. Atari, Apple II and C64. For them, the current release is completely worthless. The clear instruction for all devs is to get the cc65 HEAD, build it and go from there.

From my experience, this is true for most (larger) projects using cc65. Often those are what drives cc65 development even.

**Comment by mrdudz** [comment]
@jordigh PING! :) Do you still want to look at this?

**Comment by krischik** [comment]
> @mrdudz I am using packages that either use "head" or some pinned version all the time - official packages from my distro that

Do they compile from source? MacPorts is compiles from source. Not a good idea to have the end user compile from HEAD. 

> [@jordigh](https://github.com/jordigh) PING! :) Do you still want to look at this?

I'm still looking at it but I habe lost hope. Still quick check:

# macOS Brew:

```
>brew info cc65  
==> cc65: stable 2.19 (bottled), HEAD
6502 C compiler
https://cc65.github.io/cc65/
Installed
/usr/local/Cellar/cc65/2.19 (481 files, 31MB) *
  Poured from bottle on 2020-11-23 at 08:44:12
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/c/cc65.rb
License: Zlib
==> Options
--HEAD
	Install HEAD version
==> Caveats
Library files have been installed to:
  /usr/local/Cellar/cc65/2.19/share/cc65
==> Downloading https://formulae.brew.sh/api/formula/cc65.json
==> Analytics
```

# macOS MacPorts:

```
>port info cc65  
cc65 @2.19_1 (lang)
Variants:             docs, universal

Description:          cc65 is a complete cross development package for 65(C)02 systems, including a powerful macro assembler, a C
                      compiler, linker, librarian and several other tools.
Homepage:             https://cc65.github.io

Conflicts with:       grc
Platforms:            darwin
License:              zlib
Maintainers:          Email: ryandesign@macports.org, GitHub: ryandesign
                      Policy: openmaintainer
```

**Comment by mrdudz** [comment]
It might be a good idea to rename rgc to grc65 indeed....

However, we have not been providing macOS binaries before - so this shouldn't be a roadblock in making that GHA :)

**Comment by krischik** [comment]
> However, we have not been providing macOS binaries before - so this shouldn't be a roadblock in making that GHA :)

MacPosts doesn't provide binaries. It provides build instructions to build from source. Which has advantages and disadvantages. Brew provides binaries. That's the main difference between the two.

Both of course need a version number for the package. Sure one can make one up but that will lead to chaos.

**Comment by mrdudz** [comment]
> > However, we have not been providing macOS binaries before - so this shouldn't be a roadblock in making that GHA :)
> 
> MacPosts doesn't provide binaries. It provides build instructions to build from source. Which has advantages and disadvantages. Brew provides binaries. That's the main difference between the two.
> 
> Both of course need a version number for the package. Sure one can make one up but that will lead to chaos.

Again: This isnt about macOS. It is irrelevant to the Problem. We need a GHA that produces the Windows builds and updates the Docs etc pp

**Comment by krischik** [comment]
> We need a GHA that produces the Windows builds and updates the Docs etc pp

Why would you **need** a git hub action? You could just as well build locally and upload to github. You do that for SourceForge once a month anyway. Only difference: instead of "snapshot" add an actual version number and source zip.

You seem be stuck in what is called the nirvana fallacy: Rejecting any solution that is not absolutely perfect in all conceivable ways. However since the perfect solution is not archivable you have no solution at all.

That's, as the name suggest, is a fallacy. There are other simpler, workable and archivable solution apart from GHA.

**Comment by mrdudz** [comment]
> You do that for SourceForge once a month anyway. 

i don't do anything like that, its all done automatically.

And i am not going through another iteration of the discussion. If "releases" are desired, they also need to be automatic.

**Comment by jordigh** [comment]
Sup, remember me?

I came back with #2988

**Comment by mrdudz** [comment]
Good start, thanks - i'll look closer in a couple days or so. 

In the meantime, could you explain what it already does, what might be missing (eg does it push the releases to where they are linked from the main page? Does it update the online docs?) - and most interesting (for me), how to test it without messing up the main repo that is :)

**Comment by jordigh** [comment]
Testing without messing up the main repo is kind of tough. It's actually hardcoded to only work on the main repo. I had to apply this diff so I could test on my fork. 

```diff
diff --git a/.github/workflows/release-from-snapshot.yml b/.github/workflows/release-from-snapshot.yml
--- a/.github/workflows/release-from-snapshot.yml
+++ b/.github/workflows/release-from-snapshot.yml
@@ -16,7 +16,7 @@ concurrency:
 
 jobs:
   release:
-    if: github.repository == 'cc65/cc65'
+    if: github.repository == 'jordigh/cc65'
     runs-on: ubuntu-latest
     env:
       GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
diff --git a/.github/workflows/snapshot-on-push-master.yml b/.github/workflows/snapshot-on-push-master.yml
--- a/.github/workflows/snapshot-on-push-master.yml
+++ b/.github/workflows/snapshot-on-push-master.yml
@@ -2,7 +2,7 @@ name: Snapshot Build
 on:
   push:
     branches:
-      master
+      release-from-snapshot
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
   cancel-in-progress: true
@@ -10,7 +10,7 @@ concurrency:
 jobs:
   build_windows:
     name: Build (Windows)
-    if: github.repository == 'cc65/cc65'
+    if: github.repository == 'jordigh/cc65'
     runs-on: windows-latest
 
     steps:
@@ -31,7 +31,7 @@ jobs:
 
   build_linux:
     name: Build, Test, and Snapshot (Linux)
-    if: github.repository == 'cc65/cc65'
+    if: github.repository == 'jordigh/cc65'
     runs-on: ubuntu-latest
 
     steps:
@@ -101,27 +101,6 @@ jobs:
           path: cc65-snapshot-win64.zip
           retention-days: 90
 
-      - name: Get the online documents repo.
-        uses: actions/checkout@v4
-        with:
-          repository: cc65/doc
-          # this token will expire, if it does, generate a new one as decribed in https://github.com/cc65/cc65/issues/2065
-          # - apparently only a "classic" token works here
-          # - the token must exist in the cc65/cc65 repo
-          token: ${{ secrets.DOC_PAT }} # use secret token instead of default
-          path: doc.git
-      - name: Update the online documents.
-        run: |
-          cd doc.git
-          rm *.*
-          cp ../html/*.* .
-          git config user.name "cc65-github"
-          git config user.email "cc65.nomail@github.com"
-          git config push.default simple
-          git add -A
-          # prevent failure when there is nothing to commit
-          git diff-index --quiet HEAD || git commit -m "Updated from https://github.com/cc65/cc65/commit/${GITHUB_SHA}"
-          git push
       - name: Package offline documents.
         run: 7z a cc65-snapshot-docs.zip ./html/*.*
       - name: Upload a Documents Snapshot Zip
@@ -131,33 +110,3 @@ jobs:
           path: cc65-snapshot-docs.zip
           retention-days: 90
 
-    # enter secrets under "repository secrets"
-      - name: Upload 32-bit Windows snapshot to sourceforge
-        uses: nogsantos/scp-deploy@master
-        with:
-          src: cc65-snapshot-win32.zip
-          host: ${{ secrets.SSH_HOST }}
-          remote: ${{ secrets.SSH_DIR }}
-          port: ${{ secrets.SSH_PORT }}
-          user: ${{ secrets.SSH_USER }}
-          key: ${{ secrets.SSH_KEY }}
-      - name: Upload 64-bit Windows snapshot to sourceforge
-        uses: nogsantos/scp-deploy@master
-        with:
-          src: cc65-snapshot-win64.zip
-          host: ${{ secrets.SSH_HOST }}
-          remote: ${{ secrets.SSH_DIR }}
-          port: ${{ secrets.SSH_PORT }}
-          user: ${{ secrets.SSH_USER }}
-          key: ${{ secrets.SSH_KEY }}
-      - name: Upload documents snapshot to sourceforge
-        uses: nogsantos/scp-deploy@master
-        with:
-          src: cc65-snapshot-docs.zip
-          host: ${{ secrets.SSH_HOST }}
-          remote: ${{ secrets.SSH_DIR }}
-          port: ${{ secrets.SSH_PORT }}
-          user: ${{ secrets.SSH_USER }}
-          key: ${{ secrets.SSH_KEY }}
-
-      # TODO: Publish snapshot zip at https://github.com/cc65/cc65.github.io

```

It changes the allowed repo to be my fork and it removes the steps that requires external secrets, such as sourceforge, that I don't have access to.

"Releases" on the main page and updating docs is actually already done each time you merge to `master`. That's what [the existing `snapshot-on-push-master.yml` workflow](https://github.com/cc65/cc65/blob/master/.github/workflows/snapshot-on-push-master.yml) does.. All this does action does is create a new release in Github on the [releases page](https://github.com/cc65/cc65/releases) , using the snapshots that are already serving as *de facto* releases.

**Comment by krischik** [comment]
So is there hope we finally get an official version?

**Comment by mrdudz** [comment]
lets go on in #2988
